Fix Animation Property Types (TS2339)

# Issue #856: Fix Animation Property Types (TS2339)

## Priority
🟢 **Medium**

## Summary
Fix "Property does not exist" errors (TS2339) in accessibility test files related to jest-axe matcher.

## Current Status
- **Total TS2339 Errors**: 4
- **Affected Files**: 4 accessibility test files
- **Timeline**: 2-3 days

## Problem Description

All 4 errors are the same issue: TypeScript doesn't recognize the `toHaveNoViolations` matcher from jest-axe.

```typescript
error TS2339: Property 'toHaveNoViolations' does not exist on type 'Assertion<any>'.
```

This is a **type declaration issue**, not a runtime issue. The tests work correctly, but TypeScript doesn't know about the custom matcher.

## Errors to Fix

### 1. apple-action-sheet.a11y.test.tsx
**File**: `src/components/ui/apple-action-sheet.a11y.test.tsx:40`

**Error**:
```typescript
error TS2339: Property 'toHaveNoViolations' does not exist on type 'Assertion<any>'.
```

**Current Code**:
```typescript
const results = await axe(container);
expect(results).toHaveNoViolations();  // ❌ TypeScript error
```

---

### 2. apple-control-center.a11y.test.tsx
**File**: `src/components/ui/apple-control-center.a11y.test.tsx:39`

**Error**:
```typescript
error TS2339: Property 'toHaveNoViolations' does not exist on type 'Assertion<any>'.
```

**Current Code**:
```typescript
const results = await axe(container);
expect(results).toHaveNoViolations();  // ❌ TypeScript error
```

---

### 3. apple-live-activity.a11y.test.tsx
**File**: `src/components/ui/apple-live-activity.a11y.test.tsx:40`

**Error**:
```typescript
error TS2339: Property 'toHaveNoViolations' does not exist on type 'Assertion<any>'.
```

**Current Code**:
```typescript
const results = await axe(container);
expect(results).toHaveNoViolations();  // ❌ TypeScript error
```

---

### 4. apple-picker.a11y.test.tsx
**File**: `src/components/ui/apple-picker.a11y.test.tsx:31`

**Error**:
```typescript
error TS2339: Property 'toHaveNoViolations' does not exist on type 'Assertion<any>'.
```

**Current Code**:
```typescript
const results = await axe(container);
expect(results).toHaveNoViolations();  // ❌ TypeScript error
```

---

## Root Cause Analysis

The `toHaveNoViolations` matcher is provided by `jest-axe`, but TypeScript doesn't know about it because:

1. **Missing type declarations**: jest-axe types not properly imported
2. **Missing setup file**: Test setup file not extending Vitest matchers
3. **Incorrect import**: Not importing `toHaveNoViolations` in test files

## Solution Options

### Option A: Add Type Declaration File (RECOMMENDED)

Create a type declaration file to extend Vitest matchers with jest-axe types.

**Steps**:

1. Create `src/test-setup.d.ts`:
```typescript
import type { JestAxeConfigureOptions } from 'jest-axe';

declare global {
  namespace Vi {
    interface Assertion {
      toHaveNoViolations(): void;
    }
    interface AsymmetricMatchersContaining {
      toHaveNoViolations(): void;
    }
  }
}

export {};
```

2. Update `tsconfig.json` to include the file:
```json
{
  "include": ["src", "tests", "*.ts", "*.js", "scripts", "src/test-setup.d.ts"]
}
```

3. Ensure test setup imports jest-axe:
```typescript
// src/test-setup.ts or tests/setup.ts
import { toHaveNoViolations } from 'jest-axe';
import { expect } from 'vitest';

expect.extend(toHaveNoViolations);
```

**Pros**:
- ✅ Fixes all 4 errors at once
- ✅ Proper TypeScript solution
- ✅ No changes to test files needed
- ✅ Reusable for future a11y tests

**Cons**:
- ⚠️ Requires understanding of TypeScript declaration files

---

### Option B: Use Type Assertion in Each Test

Cast `expect` to `any` to bypass TypeScript checking.

**Current Code**:
```typescript
expect(results).toHaveNoViolations();  // ❌ TypeScript error
```

**Fixed Code**:
```typescript
(expect(results) as any).toHaveNoViolations();  // ✅ Works but not ideal
```

**Pros**:
- ✅ Quick fix
- ✅ No new files needed

**Cons**:
- ❌ Loses type safety
- ❌ Need to change 4 files
- ❌ Not a proper solution
- ❌ Goes against Issue #850 goal (remove `as any`)

---

### Option C: Import and Extend Matcher in Each Test

Import and extend matcher at the top of each test file.

**Code**:
```typescript
import { toHaveNoViolations } from 'jest-axe';
import { expect } from 'vitest';

expect.extend(toHaveNoViolations);

// Now TypeScript knows about the matcher
const results = await axe(container);
expect(results).toHaveNoViolations();  // ✅ Works
```

**Pros**:
- ✅ Explicit imports
- ✅ Type-safe

**Cons**:
- ⚠️ Need to add to each test file
- ⚠️ Repetitive

---

## Recommended Approach

**Use Option A (Type Declaration File)** because:
1. Fixes all 4 errors with one file
2. Proper TypeScript solution
3. Reusable for future tests
4. No changes to test files
5. Maintains type safety

---

## Implementation Steps

### Step 1: Create Type Declaration File

Create `src/test-setup.d.ts`:
```typescript
/// <reference types="vitest" />
import type { AxeResults } from 'axe-core';

declare global {
  namespace Vi {
    interface Assertion<T = any> {
      /**
       * Custom matcher from jest-axe for accessibility testing
       * @see https://github.com/nickcolley/jest-axe
       */
      toHaveNoViolations(): T;
    }
    
    interface AsymmetricMatchersContaining {
      toHaveNoViolations(): any;
    }
  }
}

export {};
```

### Step 2: Verify Test Setup File Exists

Check if `src/test-setup.ts` or `tests/setup.ts` exists and includes:
```typescript
import { toHaveNoViolations } from 'jest-axe';
import { expect } from 'vitest';

// Extend Vitest matchers with jest-axe
expect.extend(toHaveNoViolations);
```

If it doesn't exist, create it.

### Step 3: Update Vitest Config

Ensure `vite.config.ts` or `vitest.config.ts` includes the setup file:
```typescript
export default defineConfig({
  test: {
    setupFiles: ['./src/test-setup.ts'],
    // ... other config
  }
});
```

### Step 4: Update tsconfig.json

Ensure the type declaration file is included:
```json
{
  "include": ["src", "tests", "*.ts", "*.js", "scripts"]
  // src/test-setup.d.ts will be automatically included since it's in src/
}
```

### Step 5: Verify Fix

```bash
cd handoff/20250928/40_App/frontend-dashboard
pnpm run typecheck 2>&1 | grep "toHaveNoViolations"
# Expected: (empty output)
```

---

## Acceptance Criteria

### Must Have
- [ ] All 4 TS2339 errors fixed
- [ ] Type declaration file created
- [ ] Test setup file properly configured
- [ ] All 429 tests passing
- [ ] No new TypeScript errors introduced

### Nice to Have
- [ ] Add JSDoc comments to type declarations
- [ ] Document accessibility testing setup in README
- [ ] Add more jest-axe matchers if available

---

## Testing Plan

### 1. Type Check
```bash
pnpm run typecheck 2>&1 | grep "error TS2339"
# Expected: 0 errors (down from 4)
```

### 2. Run Accessibility Tests
```bash
pnpm test -- apple-action-sheet.a11y.test
pnpm test -- apple-control-center.a11y.test
pnpm test -- apple-live-activity.a11y.test
pnpm test -- apple-picker.a11y.test
# Expected: All tests pass
```

### 3. Verify Matcher Works
```bash
pnpm test
# Expected: All 429 tests passing, including a11y tests
```

---

## Timeline
- **Effort**: 2-3 days
  - Day 1: Create type declaration file and test setup
  - Day 2: Verify and test
  - Day 3: Documentation
- **Priority**: 🟢 Medium
- **Blocker**: None (can proceed immediately)

---

## Dependencies
- None (can proceed in parallel with other issues)

---

## Risks & Mitigation

### Risk 1: Type declarations don't work with Vitest
**Mitigation**: 
- Test with simple example first
- Check Vitest documentation for custom matchers
- Fall back to Option C if needed

### Risk 2: Tests fail after changes
**Mitigation**: 
- Run tests before and after changes
- Verify jest-axe is properly installed
- Check test setup is loaded

---

## Documentation Updates

### README.md
Add section on accessibility testing:

```markdown
## Accessibility Testing

We use [jest-axe](https://github.com/nickcolley/jest-axe) for automated accessibility testing.

### Running A11y Tests

\`\`\`bash
# Run all accessibility tests
pnpm test -- a11y.test

# Run specific component a11y test
pnpm test -- apple-button.a11y.test
\`\`\`

### Writing A11y Tests

\`\`\`typescript
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { MyComponent } from './MyComponent';

describe('MyComponent - Accessibility', () => {
  it('should have no accessibility violations', async () => {
    const { container } = render(<MyComponent />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
\`\`\`

The \`toHaveNoViolations\` matcher is automatically available via our test setup.
```

---

## Related Issues
- Part of Milestone: "TypeScript Cleanup 119→0"
- Related: Issue #851 (Storybook type safety)
- Can proceed in parallel with: Issue #854, #855, #857

---

## Labels
- `ts-error:TS2339`
- `area:tests`
- `area:accessibility`
- `priority:medium`
- `type:bug`

---

**Created**: 2025-10-27  
**Estimated Effort**: 2-3 days  
**Impact**: -4 errors  
**Status**: Ready to implement

--- a/handoff/20250928/40_App/frontend-dashboard/src/test/setup.d.ts
+++ b/handoff/20250928/40_App/frontend-dashboard/src/test/setup.d.ts
@@ -1,18 +1,16 @@
-import type { AxeResults } from 'axe-core';
+/// <reference types="vitest" />
 
-declare global {
-  namespace Vi {
-    interface Assertion<T = any> {
-      /**
-       * Custom matcher from jest-axe for accessibility testing
-       * @see https://github.com/nickcolley/jest-axe
-       */
-      toHaveNoViolations(): T;
-    }
-    
-    interface AsymmetricMatchersContaining {
-      toHaveNoViolations(): any;
-    }
+declare module 'vitest' {
+  interface Assertion<T = any> {
+    /**
+     * Custom matcher from jest-axe for accessibility testing
+     * @see https://github.com/nickcolley/jest-axe
+     */
+    toHaveNoViolations(): void;
+  }
+  
+  interface AsymmetricMatchersContaining {
+    toHaveNoViolations(): void;
   }
 }
 

